### ISSUE TITLE
Problems with Lists of Dicts.

### ISSUE THREAD

**Issue by vladola** [pydantic/pydantic#2868]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.2
            pydantic compiled: True
                 install path: /Users/vlad/code/1p-tag-editor/.venv/lib/python3.9/site-packages/pydantic
               python version: 3.9.1 (default, Feb  3 2021, 07:04:48)  [Clang 11.0.0 (clang-1100.0.33.17)]
                     platform: macOS-10.14.6-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

```py
from pydantic import BaseModel
from typing import List, Dict, TypedDict, Optional
from fastapi import FastAPI

class PortResData(BaseModel):
    change: str
    user: str
    date: str

class PortWithResData(BaseModel):
    ports: List[Dict[str, PortResData]]

app = FastAPI()


@app.post("/items")
def read_item(port: PortWithResData):
    return {"port": port}

```

When going to `http://127.0.0.1:8000/docs#/default/read_item_items_post`, the "schema" is presented as:

```json
{
  "ports": [
    {
      "additionalProp1": {
        "change": "string",
        "user": "string",
        "date": "string"
      },
      "additionalProp2": {
        "change": "string",
        "user": "string",
        "date": "string"
      },
      "additionalProp3": {
        "change": "string",
        "user": "string",
        "date": "string"
      }
    }
  ]
}
```

Instead, it should be presented as:

```json
{
   "ports":[
      {
         "additionalProp1":{
            "change":"string",
            "user":"string",
            "date":"string"
         }
      },
      {
         "additionalProp2":{
            "change":"string",
            "user":"string",
            "date":"string"
         }
      },
      {
         "additionalProp3":{
            "change":"string",
            "user":"string",
            "date":"string"
         }
      }
   ]
}
```

TL;DR: Instead of `{"ports": list_of_dicts}` I'm getting `{"ports": big_nested_dict}`, so the "ports" list has only 1 item, the big nested dict instead of many self-contained dicts.

**Comment by vladola** [comment]
For anyone coming back to this: I temporarily "solved" this by manually defining an example schema like described [here]. Since it's manually declared, the schema example is correct as long as you gave a correct example.

[here]: https://fastapi.tiangolo.com/tutorial/schema-extra-example/

```py
class PortWithResData(BaseModel):
    __root__: List[Dict[str, PortResData]]

    class Config:
        schema_extra = {
            "example": 
                [
                    {"Eth1": {
                        "change": "INC001",
                        "user": "my_AD_USER",
                        "date": "2021-01-01"
                    }},
                    {"Eth2": {
                        "change": "INC001",
                        "user": "my_AD_USER",
                        "date": "2021-01-01"
                    }},
                ]
            
        }

**Comment by adriangb** [comment]
Pydantic is producing what I think is the correct schema in both V1 and V2:

```python
from typing import Dict, List

from pydantic import BaseModel


class PortResData(BaseModel):
    change: str
    user: str
    date: str


class PortWithResData(BaseModel):
    ports: List[Dict[str, PortResData]]


schema = PortWithResData.schema_json()

expected = {
    "title": "PortWithResData",
    "type": "object",
    "properties": {
        "ports": {
            "title": "Ports",
            "type": "array",
            "items": {
                "type": "object",
                "additionalProperties": {"$ref": "#/definitions/PortResData"},
            },
        }
    },
    "required": ["ports"],
    "definitions": {
        "PortResData": {
            "title": "PortResData",
            "type": "object",
            "properties": {
                "change": {"title": "Change", "type": "string"},
                "user": {"title": "User", "type": "string"},
                "date": {"title": "Date", "type": "string"},
            },
            "required": ["change", "user", "date"],
        }
    },
}

assert schema == expected
```

Here's the `openapi.json` that I'm getting, which also looks correct:

```json
{
    "openapi": "3.0.2",
    "info": {
        "title": "FastAPI",
        "version": "0.1.0"
    },
    "paths": {
        "/items": {
            "post": {
                "summary": "Read Item",
                "operationId": "read_item_items_post",
                "requestBody": {
                    "content": {
                        "application/json": {
                            "schema": {
                                "$ref": "#/components/schemas/PortWithResData"
                            }
                        }
                    },
                    "required": true
                },
                "responses": {
                    "200": {
                        "description": "Successful Response",
                        "content": {
                            "application/json": {
                                "schema": {}
                            }
                        }
                    },
                    "422": {
                        "description": "Validation Error",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/HTTPValidationError"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {
            "HTTPValidationError": {
                "title": "HTTPValidationError",
                "type": "object",
                "properties": {
                    "detail": {
                        "title": "Detail",
                        "type": "array",
                        "items": {
                            "$ref": "#/components/schemas/ValidationError"
                        }
                    }
                }
            },
            "PortResData": {
                "title": "PortResData",
                "required": [
                    "change",
                    "user",
                    "date"
                ],
                "type": "object",
                "properties": {
                    "change": {
                        "title": "Change",
                        "type": "string"
                    },
                    "user": {
                        "title": "User",
                        "type": "string"
                    },
                    "date": {
                        "title": "Date",
                        "type": "string"
                    }
                }
            },
            "PortWithResData": {
                "title": "PortWithResData",
                "required": [
                    "ports"
                ],
                "type": "object",
                "properties": {
                    "ports": {
                        "title": "Ports",
                        "type": "array",
                        "items": {
                            "type": "object",
                            "additionalProperties": {
                                "$ref": "#/components/schemas/PortResData"
                            }
                        }
                    }
                }
            },
            "ValidationError": {
                "title": "ValidationError",
                "required": [
                    "loc",
                    "msg",
                    "type"
                ],
                "type": "object",
                "properties": {
                    "loc": {
                        "title": "Location",
                        "type": "array",
                        "items": {
                            "anyOf": [
                                {
                                    "type": "string"
                                },
                                {
                                    "type": "integer"
                                }
                            ]
                        }
                    },
                    "msg": {
                        "title": "Message",
                        "type": "string"
                    },
                    "type": {
                        "title": "Error Type",
                        "type": "string"
                    }
                }
            }
        }
    }
}
```

So maybe Swagger is translating the OpenAPI schema in a way that you don't like or is incorrect?
